### PR TITLE
chore: Split functions crate into multi crates

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

- fixes: #17117

## Tests

- [ ] Unit Test
- [ ] Logic Test
- [ ] Benchmark Test
- [x] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [x] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/databendlabs/databend/17133)

### CODE REVIEW

**Comment by TCeason** [comment]
Before split:

image

In pr:

image

So I think no need to split variant, string and others...

**Comment by TCeason** [comment]
In the 069117acb79fe2547477436eff0b5a4b57aa6af8 commit :
we split databend-functions-scalar-arithmetic databend-functions-scalar-arithmetic-modulo databend-functions-scalar-datetime databend-functions-scalar-decimal-utils.

databend-functions-scalar-arithmetic is slowest. because it has more macros.
image

**Comment by TCeason** [comment]
In commit https://github.com/databendlabs/databend/pull/17133/commits/b35433d13079027c031fdff8863447ae2e38bdb6 

image

File: tests/sqllogictests/suites/mode/standalone/ee/explain_agg_index.test [hunk 1]
```diff
@@ -325,7 +325,7 @@ EXPLAIN SELECT station_name, MIN(measurement) AS min_measurement, AVG(measuremen
 ----
 EvalScalar
 ├── output columns: [MIN(measurement) (#2), MAX(measurement) (#5), onebrc.station_name (#0), mean_measurement (#6)]
-├── expressions: [sum(measurement) (#3) / CAST(if(CAST(count(measurement) (#4) = 0 AS Boolean NULL), 1, count(measurement) (#4)) AS UInt64 NULL)]
+├── expressions: [sum(measurement) (#3) / CAST(if(CAST(count(measurement) (#4) = 0 AS Boolean NULL), 1, count(measurement) (#4)) AS Float64 NULL)]
```
Comment (TCeason) [inline tests/sqllogictests/suites/mode/standalone/ee/explain_agg_index.test:328]: In this pr, we remove int/double arithmetic, so the type changed. I think it's ok.

And after we split crate, we also can contain int/double arithmetic.

Comment (sundy-li) [reply tests/sqllogictests/suites/mode/standalone/ee/explain_agg_index.test:328]: > And after we split crate, we also can contain int/double arithmetic.

Yes, Let's just split the crate and don't remove registers which keep the behavior as before.

If something incompat happens, we don't need to revert this pr. 

We can remove these after this pr.

Comment (TCeason) [reply tests/sqllogictests/suites/mode/standalone/ee/explain_agg_index.test:328]: Ok, I keep the registers function. https://github.com/databendlabs/databend/pull/17133/commits/4d189c240b023388ff368ce3ed460dfb463d08a8

image

File: src/query/functions/tests/it/scalars/testdata/function_list.txt [hunk 2]
```diff
@@ -2580,228 +2580,228 @@ Functions overloads:
...
-217 minus(Float64 NULL, Int64 NULL) :: Float64 NULL
-218 minus(Float32, Float32) :: Float64
-219 minus(Float32 NULL, Float32 NULL) :: Float64 NULL
-220 minus(Float32, Float64) :: Float64
-221 minus(Float32 NULL, Float64 NULL) :: Float64 NULL
-222 minus(Float64, Float32) :: Float64
-223 minus(Float64 NULL, Float32 NULL) :: Float64 NULL
-224 minus(Float64, Float64) :: Float64
-225 minus(Float64 NULL, Float64 NULL) :: Float64 NULL
+4 minus FACTORY
```
Comment (sundy-li) [inline src/query/functions/tests/it/scalars/testdata/function_list.txt:2583]: Something is changed here.

Comment (sundy-li) [reply src/query/functions/tests/it/scalars/testdata/function_list.txt:2583]: The function order should be consistent if we only refactor the crates by spiliting.

Comment (TCeason) [reply src/query/functions/tests/it/scalars/testdata/function_list.txt:2583]: Get

File: src/query/functions/src/scalars/decimal/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "databend-functions-scalar-decimal-utils"
```
Comment (sundy-li) [inline src/query/functions/src/scalars/decimal/Cargo.toml:2]: name should be `databend-functions-scalar-decimal`

Comment (TCeason) [reply src/query/functions/src/scalars/decimal/Cargo.toml]: Done. and keep the register order.

image

**Review by sundy-li — APPROVED** [review]

**Pull request merged**